### PR TITLE
add tests for nan in data with zero weights causing issues in mean, var, std, cov

### LINKED ISSUE #111: `nan` handling of statistical functions such as `mean`, `std` etc.
Currently the statistical functions such as `mean`, `std` etc. return `nan` if there are `nan` in the data, even if the corresponding weights are zero. 

To fix this, the zero weight data should be discarded before calculating the mean etc., similarly to #102, where contour plots were fixed for `nan` in the data.

### PR DESCRIPTION
**Pull request by lukashergt**
Make a change similar to #102 for statistical functions (mean, var, etc.) in `WeightedSeries` and `WeightedDataFrame` to better handle `nan` in the data.

Fixes #111 

# Checklist:

- [x] I have performed a self-review of my own code
- [x] My code is PEP8 compliant (`flake8 anesthetic tests`)
- [x] My code contains compliant docstrings (`pydocstyle --convention=numpy anesthetic`)
- [x] New and existing unit tests pass locally with my changes (`python -m pytest`)
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: anesthetic/weighted_pandas.py [hunk 1]
```diff
@@ -55,11 +55,18 @@ def __init__(self, *args, **kwargs):
 
     def mean(self):
         """Weighted mean of the sampled distribution."""
-        return np.average(self, weights=self.weight)
+        if self.weight is None:
+            return np.average(self)
+        nonzero = self.weight > 0
+        return np.average(self[nonzero], weights=self.weight[nonzero])
```
Comment (lukashergt) [inline anesthetic/weighted_pandas.py:61]: The failing code coverage comes from the `if.self.weight is None` case.

Do we need to check for this? Or are we confident that any `weight=None` is turned into `weight=1` by the weighted objects?

**Comment by lukashergt** [comment]
> The more 'pandas' way to do this would probably be
> 
> ```
> samples.dropna().mean()
> ```
> 
> or
> 
> ```
> samples.mean(skipna=True)
> ```

I am unsure whether we would want to fully ignore `nan`. This way we only discard `nan` if they have zero weight, which is the safer way, I think.

**Comment by lukashergt** [comment]
> At the moment we don't have negative weights, but is there any reason we need to exclude them here? It would make more sense in terms of readability for `nonzero = self.weight !=0`

Are negative weights a thing? But true, we actually only want to discard zero weights. 

I'll change it accordingly.

File: anesthetic/weighted_pandas.py (hunk 1 above)
Comment (lukashergt) [reply anesthetic/weighted_pandas.py:61]: Tests run fine upon deleting those lines...

**Comment by williamjameshandley** [comment]
> I am unsure whether we would want to fully ignore `nan`. This way we only discard `nan` if they have zero weight, which is the safer way, I think.

I agree. This is why I think that your solution of dropping zero weight nans, and then leaving it up to the user if they want to drop all nans is the correct thing to do. It's important for future developers (including us) that we record this here as the reasoning for doing so.

File: anesthetic/weighted_pandas.py (hunk 1 above)
Comment (williamjameshandley) [reply anesthetic/weighted_pandas.py:61]: I think that this comes from an earlier version of the code when weights were treated slightly differently. `weight` should now never be `None` (if `_weight==None`, then `weight` is returned as a set of appropriate `1.`s). We should drop it.

**Comment by lukashergt** [comment]
> but `skipna` would be a good arg to add in this pull request, so that if you want you can also easily skip non-zero weight nans.

According to [pandas](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.mean.html) the signature of `skipna` is `None` but the default is `True`, is that what we want, too? 

I'm not sure whether skipping `nan` with nonzero weight is indeed what we want...

**Comment by williamjameshandley** [comment]
Hmm. I agree, although it does mean that NestedSamples has different behaviour with repect to pandas. Leave the PR as it is for now.
